File: src/DoublyLinkedList.java

public class DoublyLinkedList {
	private Node head;
	private Node tail;
	private int length = 0;
	
	class Node {
		int value;
		Node prev;
		Node next;
		
		Node(int value) {
			this.value = value;
		}
	}
	
	DoublyLinkedList(int value) {
		Node node = new Node(value);
		head = node;
		tail = node;
		length = 1;
	}
	
	// getHead(), getTail(), getLength()
	
	public void getHead() {
		if (head == null) {
			System.out.println("DoublyLinkedList Empty!");
		} else {
			System.out.println("HEAD: " + head.value);
		}
	}
	
	public void getTail() {
		if (tail == null) {
			System.out.println("DoublyLinkedList Empty!");
		} else {
			System.out.println("TAIL: " + tail.value);
		}
	}
	
	public void getLength() {
		System.out.println("Length: " + length);
	}
	
	// append(value) 
	
	public void append(int value)  {
		Node node = new Node(value);
		if (length == 0) {
			head = node;
			tail = node;
		} else {
			tail.next = node;
			node.prev = tail;
			tail = node;
		}
		length ++;
	}
	
	// removeLast
	
	public Node removeLast() {
		if (length == 0) return null;
		Node temp = tail;
		if (length == 1) {
			head = null;
			tail = null;
		} else {
			tail = tail.prev;
			tail.next = null;
			temp.prev = null;
		}
		length --;
		return temp;
	}
	
	// 	prepend(value)
	
	public void prepend(int value) {
		Node node = new Node(value);
		if(length == 0) {
			head = node;
			tail = node;
		} else {
			node.next = head;
			head.prev = node;
			head = node;
		}
		length ++;
	}
	
	// 	removeFirst
	
	public Node removeFirst() {
		if (length == 0) return null;
		Node temp = head;
		if (length == 1) {
			head = null;
			tail = null;
		} else {
			head = head.next;
			head.prev = null;
			temp.next = null;
		}
		length --;
		return temp;
	}
	
	// get(int index)
	
	public Node get(int index) {
		if (index < 0 || index >= length) return null;
		Node temp = head;
		if (index < length / 2) {
			int ind = 0;
			while(ind != index) {
				ind ++;
				temp = temp.next;
			}
		} else {
			temp = tail;
			int ind = length - 1;
			while(ind != index) {
				ind --;
				temp = temp.prev;
			}
		}
		return temp;
	}
	
	// set(int index, int value)
	
	public boolean set(int index, int value) {
		Node node = get(index);
		if (node == null) return false;
		node.value = value;
		return true;
	}
	
	// insert to any index.
	
	public boolean insert(int index, int value) {
		if (index < 0 || index > length) return false;
		if (index == 0) {
			prepend(value);
			return true;
		}
		if (index == length) {
			append(value);
			return true;
		}
		Node node = new Node(value);
		Node before = get(index - 1);
		Node after = get(index);
		node.next = after;
		node.prev = before;
		after.prev = node;
		before.next = node;
		length ++;
		return true;
	}
	
	// remove at any index 
	
	public Node remove(int index) {
		if (index < 0 || index >= length) return null;
		if (index == 0) return removeFirst();
		if (index == length - 1) return removeLast();
		Node temp = get(index);
		Node before = temp.prev;
		Node after = temp.next;
		before.next = after;
		after.prev = before;
		temp.prev = null;
		temp.next = null;
		length --;
		return temp;
	}
	
	public void printList() {
		Node node = head;
		while (node != null) {
			System.out.println(node.value);
			node = node.next;
		}
	}


}

File: Javascript/DS/doublylinkedlist/DoublyLinkedList.js
class Node {
    
    constructor (data) {
        this.data = data
        this.prev = null
        this.next = null
    }

}

class DoublyLinkedList {

    constructor() {
        this.head = null
        this.tail = null
        this.length = 0
    }

    print() {
        let iterator = this.head
        console.log("---DLL---")
        while (iterator !== null) {
            console.log(iterator.data)
            iterator = iterator.next
        }
        console.log("---END---")
    }

    add(data) {
        let elementNode = new Node(data)
        if (this.length === 0) {
            this.head = elementNode
            this.tail = elementNode
            this.length ++
            return
        }
        elementNode.prev = this.tail
        this.tail.next = elementNode
        this.tail = elementNode
        this.length ++
    }

    remove(data) {
        if (this.length === 0) return null
        let iterator = this.head
        while (iterator && this.length !== 0) {
            if (iterator.data === data) {
                if (this.length === 1) {
                    this.head = null
                    this.tail = null
                    this.length = 0
                    return
                }
                if (iterator === this.head) {
                    this.head = this.head.next
                    this.head.prev = null
                    this.length --
                } else if (iterator === this.tail) {
                    this.tail = this.tail.prev
                    this.tail.next = null
                    this.length --
                } else {
                    const tempNext = iterator.next
                    iterator.prev.next = iterator.next
                    tempNext.prev = iterator.prev
                    this.length --
                }
                
            }
            iterator = iterator.next
        }
        
    }

    reverse() {
        if (this.length === 0) return null
        var before = null
        var after = this.head
        var current = this.head
        this.tail = current
        while (current) {
            after = after.next
            current.next = before
            current.prev = after
            before = current
            current = after
        }
        this.head = before
    }

}

module.exports = DoublyLinkedList